Issue 52190: Expose Postgres connections and locks to admins

labkey-jeckels opened the pull request:
#### Rationale
To troubleshoot deadlocks and performance issues it's often useful to get the status of the database connections, what they might be blocked on, and what locks are involved. By making this available through the app instead of strictly though Postgres tools we can make this faster and easier for admins.

#### Changes
- Expose new `pg_stat_activity` and `pg_locks` queries in the root container in a new `postgres` schema, only available to App/Site Admins and Troubleshooters in the root container
- Links to those queries from the Admin Console
- Support killing Postgres connections via "delete" on `postgresConnections`
- Include the results of the queries when exporting server diagnostics
- Show Java threads and HTTP requests associated with DB connections

labkey-jeckels committed:
Issue 52190: Expose Postgres connections and locks to admins

labkey-nicka commented on core/src/org/labkey/core/query/PostgresConnectionsTable.java:
```diff
@@ -0,0 +1,137 @@
+
+import java.util.Map;
+
+public class PostgresConnectionsTable extends VirtualTable<CoreQuerySchema>
```
Making select rows requests against this table fail if the 17.1 API version is utilized:

```js
LABKEY.Query.selectRows({
    containerPath: '/',
    schemaName: 'core',
    queryName: 'PostgresConnections',
    requiredVersion: 17.1,
    success: () => {},
    failure: () => {},
});
```

causes:

```
WARN  ApiResponseWriter        2025-02-10T14:30:29,522     http-nio-8080-exec-4 : ApiResponseWriter exception: 
com.fasterxml.jackson.databind.exc.InvalidDefinitionException: No serializer found for class org.postgresql.core.v3.replication.V3ReplicationProtocol and no properties discovered to create BeanSerializer (to avoid exception, disable SerializationFeature.FAIL_ON_EMPTY_BEANS) (through reference chain: org.postgresql.jdbc.PgArray["resultSet"]->org.postgresql.jdbc.PgResultSet["statement"]->org.postgresql.jdbc.PgStatement["connection"]->org.postgresql.jdbc.PgConnection["queryExecutor"]->org.postgresql.core.v3.QueryExecutorImpl["replicationProtocol"])
	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:77) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.SerializerProvider.reportBadDefinition(SerializerProvider.java:1328) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.DatabindContext.reportBadDefinition(DatabindContext.java:414) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.impl.UnknownSerializer.failForEmpty(UnknownSerializer.java:53) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.impl.UnknownSerializer.serialize(UnknownSerializer.java:30) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:732) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:770) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:184) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:732) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:770) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:184) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:732) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:770) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:184) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:732) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:770) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:184) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.BeanPropertyWriter.serializeAsField(BeanPropertyWriter.java:732) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFields(BeanSerializerBase.java:770) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:184) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider._serialize(DefaultSerializerProvider.java:502) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:341) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.databind.ObjectMapper.writeValue(ObjectMapper.java:3365) ~[jackson-databind-2.18.2.jar:2.18.2]
	at com.fasterxml.jackson.core.base.GeneratorBase.writeObject(GeneratorBase.java:422) ~[jackson-core-2.18.2.jar:2.18.2]
	at org.labkey.api.action.ApiJsonWriter.writeObject(ApiJsonWriter.java:234) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiJsonWriter.writeProperty(ApiJsonWriter.java:169) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiJsonWriter.writeProperties(ApiJsonWriter.java:284) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiJsonWriter.writeMap(ApiJsonWriter.java:271) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiJsonWriter.writeObject(ApiJsonWriter.java:182) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiJsonWriter.writeProperty(ApiJsonWriter.java:169) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiJsonWriter.writeProperties(ApiJsonWriter.java:284) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiJsonWriter.writeMap(ApiJsonWriter.java:271) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiJsonWriter.writeObject(ApiJsonWriter.java:182) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiJsonWriter.writeProperty(ApiJsonWriter.java:169) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiJsonWriter.writeProperties(ApiJsonWriter.java:284) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiJsonWriter.writeMap(ApiJsonWriter.java:271) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiJsonWriter.writeObject(ApiJsonWriter.java:182) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiJsonWriter.writeListEntry(ApiJsonWriter.java:318) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiQueryResponse.writeRowset(ApiQueryResponse.java:559) ~[api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiQueryResponse.render(ApiQueryResponse.java:160) [api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ReportingApiQueryResponse.render(ReportingApiQueryResponse.java:69) [api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiResponseWriter.write(ApiResponseWriter.java:328) [api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ApiResponseWriter.writeResponse(ApiResponseWriter.java:249) [api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.BaseApiAction.writeResponse(BaseApiAction.java:148) [api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.BaseApiAction.handlePost(BaseApiAction.java:247) [api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ReadOnlyApiAction.handleGet(ReadOnlyApiAction.java:46) [api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.BaseApiAction.handleRequest(BaseApiAction.java:131) [api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.BaseViewAction.handleRequest(BaseViewAction.java:195) [api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.ReadOnlyApiAction.handleRequest(ReadOnlyApiAction.java:40) [api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.action.SpringActionController.handleRequest(SpringActionController.java:527) [api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.module.DefaultModule.dispatch(DefaultModule.java:1124) [api-25.3-SNAPSHOT.jar:?]
	at org.labkey.api.view.ViewServlet._service(ViewServlet.java:202) [api-25.3-SNAPSHOT.jar:?]
```

labkey-jeckels replied:
Can you retest after my changes? It's not reproing for me after I changed the column types (casting the `blocked_by` array to varchar and using milliseconds instead of an interval)

labkey-nicka replied:
Verified this no longer occurs.

labkey-nicka commented on core/src/org/labkey/core/query/PostgresConnectionsTable.java:
```diff
@@ -0,0 +1,137 @@
+        addColumn(new ExprColumn(this, "wait_event_type", new SQLFragment(ExprColumn.STR_TABLE_ALIAS + ".wait_event_type"), JdbcType.VARCHAR));
+        addColumn(new ExprColumn(this, "wait_event", new SQLFragment(ExprColumn.STR_TABLE_ALIAS + ".wait_event"), JdbcType.VARCHAR));
+        addColumn(new ExprColumn(this, "query_start", new SQLFragment(ExprColumn.STR_TABLE_ALIAS + ".query_start"), JdbcType.TIMESTAMP));
+        addColumn(new ExprColumn(this, "running_time", new SQLFragment(ExprColumn.STR_TABLE_ALIAS + ".running_time"), JdbcType.OTHER));
```
The running time is not displaying as I would've expected:

<img width="1161" alt="image" src="https://github.com/user-attachments/assets/4d430f50-0926-42e4-9c7d-3a0fdd03f1af" />

It seems OK but we tend to display other query information in ms. Should we consider having this in ms? This is also difficult to parse as a client response:

```json
{
    "wait_event" : "ClientRead",
    "blocked_by" : "{}",
    "wait_event_type" : "Client",
    "query" : "-- <QueryServiceImpl .... \n",
    "query_start" : "2025-02-10 14:22:13.720",
    "running_time" : "0 years 0 mons 0 days 0 hours 16 mins 43.529306 secs",
    "pid" : 7728,
    "usename" : "postgres",
    "state" : "idle"
  }

labkey-jeckels replied:
Switched to an integer for ms

labkey-nicka replied:
This is much more straightforward. Also, thanks for updating the query to account for idle vs active queries.

labkey-nicka commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -463,6 +464,13 @@ public static void registerAdminConsoleLinks()
+
+        if (CoreSchema.getInstance().getSqlDialect().isPostgreSQL())
+        {
+            AdminConsole.addLink(Diagnostics, "postgres connections", PageFlowUtil.urlProvider(QueryUrls.class).urlExecuteQuery(ContainerManager.getRoot(), CoreQuerySchema.NAME, CoreQuerySchema.POSTGRES_CONNECTIONS_TABLE_NAME));
```
Our admin console pages tend to link back to the "Admin Console" in the navigation. These pages link back to the schema browser. I realize here we're linking over the the generic query view. Calling it out here as an inconsistency:

<img width="363" alt="image" src="https://github.com/user-attachments/assets/09e900ab-52b2-44a8-8afd-3b5f141a5bad" />

labkey-jeckels replied:
Yes. I could introduce a dedicated action but not sure it's worth it.

labkey-nicka replied:
I'd say it is worth it since this is the only view that I see that doesn't navigate back to the admin console. We made a concerted effort to normalize these pages so that they link back consistently. Here is an [example issue](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=51099).

labkey-jeckels replied:
Updated

labkey-nicka replied:
Works great and still allows for correct navigation linkage when coming from the schema browser. Thanks!

labkey-nicka commented on core/src/org/labkey/core/query/PostgresLocksTable.java:
```diff
@@ -0,0 +1,50 @@
+{
+    public PostgresLocksTable(@Nullable CoreQuerySchema userSchema)
+    {
+        super(userSchema.getDbSchema(), CoreQuerySchema.POSTGRES_LOCKS_TABLE_NAME, userSchema);
```
The `userSchema` can be null but then is immediately referenced here.

labkey-jeckels replied:
Not sure how these ended up Nullable. Switched to NotNull

labkey-nicka commented on core/src/org/labkey/core/query/PostgresLocksTable.java:
```diff
@@ -0,0 +1,50 @@
+        setDescription("Shows info about the currently held Postgres locks");
+
+        // https://www.postgresql.org/docs/current/view-pg-locks.html
+        addColumn(new ExprColumn(this, "locktype", new SQLFragment(ExprColumn.STR_TABLE_ALIAS + ".locktype"), JdbcType.VARCHAR));
```
nit: This all seems a bit repetitive in the case where we're just echoing the table through and we have access to the table metadata. Easier said than done though maybe.

labkey-jeckels replied:
Agreed!

labkey-nicka commented on core/src/org/labkey/core/query/PostgresConnectionsTable.java:
```diff
@@ -0,0 +1,137 @@
+        @Override
+        protected Map<String, Object> getRow(User user, Container container, Map<String, Object> keys)
+        {
+            return new TableSelector(PostgresConnectionsTable.this).getMap(getPid(keys));
```
Should this have the same `pid == null` protection check like `deleteRow()` has?

labkey-jeckels replied:
`getMap()` will return null. I'll add a comment to that effect

labkey-nicka commented on core/src/org/labkey/core/query/PostgresConnectionsTable.java:
```diff
@@ -0,0 +1,137 @@
+    public @NotNull SQLFragment getFromSQL()
+    {
+        SQLFragment result = new SQLFragment();
+        result.append("""
```
I realize in the issue I listed a specific query but is there utility in opening this up to the whole [set of columns available in pg_stat_activity](https://www.postgresql.org/docs/current/monitoring-stats.html#MONITORING-PG-STAT-ACTIVITY-VIEW)? That said it is is nice to have the calculated columns `blocked_by` and `running_time`.

No action necessary here just thinking.

labkey-jeckels replied:
Switched to `*` plus the calculated columns, and tried to choose a useful set of defaults.

labkey-nicka commented on core/src/org/labkey/core/query/PostgresConnectionsTable.java:
```diff
@@ -0,0 +1,137 @@
+
+        setDescription("Shows info about the active Postgres connections and their activity");
+
+        addColumn(new ExprColumn(this, "usename", new SQLFragment(ExprColumn.STR_TABLE_ALIAS + ".usename"), JdbcType.VARCHAR));
```
nit: Consider adding a similar contextual link the [pg_stat_activity docs](https://www.postgresql.org/docs/current/monitoring-stats.html#MONITORING-PG-STAT-ACTIVITY-VIEW) as you did for the locks view.

labkey-jeckels replied:
Added

labkey-nicka commented on core/src/org/labkey/core/query/PostgresConnectionsTable.java:
```diff
@@ -0,0 +1,137 @@
+        @Override
+        protected Map<String, Object> deleteRow(User user, Container container, Map<String, Object> oldRow) throws InvalidKeyException
+        {
+            Integer pid = getPid(oldRow);
```
Super powerful. This means queries can be killed via our query APIs instead of having to go through our admin pages.

labkey-jeckels replied:
Powerful and dangerous. Worth an extra review on the permission checks.

labkey-nicka commented on core/src/org/labkey/core/query/PostgresLocksTable.java:
```diff
@@ -0,0 +1,50 @@
+        addColumn(new ExprColumn(this, "mode", new SQLFragment(ExprColumn.STR_TABLE_ALIAS + ".mode"), JdbcType.VARCHAR));
+        addColumn(new ExprColumn(this, "granted", new SQLFragment(ExprColumn.STR_TABLE_ALIAS + ".granted"), JdbcType.BOOLEAN));
+        addColumn(new ExprColumn(this, "fastpath", new SQLFragment(ExprColumn.STR_TABLE_ALIAS + ".fastpath"), JdbcType.BOOLEAN));
+        addColumn(new ExprColumn(this, "waitstart", new SQLFragment(ExprColumn.STR_TABLE_ALIAS + ".waitstart"), JdbcType.TIMESTAMP));
```
It looks like some of these columns (e.g. `waitstart`) have been introduced in versions of Postgres after v13. Any way we can confirm with table metadata the presence of a column in the database before adding?

labkey-jeckels replied:
Indeed. I thought I was looking at the docs for v13 but it must have been v14 or later. Trimmed.

labkey-nicka replied:
I pulled down v13 (specifically 13.18) and performed manual testing on these. Looks good.

labkey-nicka reviewed (commented):
This will be very useful!

labkey-jeckels committed:
Code review improvements

labkey-jeckels committed:
Show PID, not display value

labkey-jeckels committed:
Add logging when PG connection is killed

labkey-nicka commented on core/src/org/labkey/core/query/PostgresConnectionsTable.java:
```diff
@@ -0,0 +1,179 @@
+{
+    private static final Logger LOG = LogHelper.getLogger(PostgresConnectionsTable.class, "Access to Postgres connection status");
+
+    public PostgresConnectionsTable(@NotNull CoreQuerySchema userSchema)
```
Throwing out an idea here. What if instead of putting these queries on the core schema we put them on a new "postgres" schema. And then exposed them as the names that they are, namely, pg_stat_activity and pg_locks? This jives since we are now doing `SELECT *` from these tables. Seems like going through our somewhat arbitrary naming pattern is something we'll just end up explaining in documentation and to people directly. We can still label them as "Postgres Connections" and "Postgres Locks".

labkey-jeckels replied:
Updated

labkey-nicka replied:
I feel like this aligns really nicely. Thanks!

labkey-jeckels committed:
Use "postgres" schema and "pg_locks"/"pg_stat_activity" as schema and query names; include thread info

labkey-jeckels committed:
Merge branch 'develop' into fb_52190_postgresConnections

labkey-jeckels committed:
Improve admin console navigation

Associate queries more closely with Postgres views

Show threads and HTTP requests associated with queries

labkey-jeckels committed:
Merge branch 'develop' into fb_52190_postgresConnections

labkey-jeckels committed:
Use constant for Postgres schema name

labkey-jeckels committed:
Merge branch 'develop' into fb_52190_postgresConnections

labkey-jeckels committed:
Merge branch 'develop' into fb_52190_postgresConnections

labkey-jeckels committed:
Automated tests for Postgres troubleshooting queries

labkey-jeckels committed:
Merge branch 'develop' into fb_52190_postgresConnections

labkey-jeckels committed:
Merge branch 'develop' into fb_52190_postgresConnections
